### PR TITLE
gh-115801: Only allow sequence of strings as input for difflib.unified_diff

### PR DESCRIPTION
**Pull request by eendebakpt**
`difflib.unified_diff` (and `difflib.context_diff `) are documented to have lists of strings as input. When called with strings as input arguments, e.g. `difflib.unified_diff('one', 'two')` a diff is generated, instead of returning an error.

In this PR we disallow input of pure strings to prevent users from accidentally passing a string to `difflib.unified_diff`.

* Two tests had to be modified, but it looks like these tests were written to test something else and not passing strings as arguments
* We still allow other sequence types such as a tuple of strings.
* This is a backwards incompatible change. If we do not want this, then I will modify the PR to update the documentation and add tests for this behavior.

* Issue: gh-115801

### CODE REVIEW

File: Lib/difflib.py [hunk 1]
```diff
@@ -1266,6 +1266,8 @@ def _check_types(a, b, *args):
     if b and not isinstance(b[0], str):
         raise TypeError('lines to compare must be str, not %s (%r)' %
                         (type(b[0]).__name__, b[0]))
+    if isinstance(a, str) or isinstance(b, str):
```
Comment (eamanu) [inline Lib/difflib.py]: Following the existing code, IMO it would be great add the `not %s (%r)'` phrase.

File: Lib/test/test_difflib.py [hunk 2]
```diff
@@ -273,9 +273,19 @@ def test_make_file_usascii_charset_with_nonascii_input(self):
         self.assertIn('&#305;mpl&#305;c&#305;t', output)
 
 
+class TestInputParsing(unittest.TestCase):
+    def test_input_type_checks(self):
+        with self.assertRaises(Exception):
+            list(difflib.unified_diff('a', ['b']))
+        with self.assertRaises(TypeError):
+            list(difflib.unified_diff(['a'], 'b'))
+        with self.assertRaises(Exception):
```
Comment (serhiy-storchaka) [inline Lib/test/test_difflib.py]: What exactly exception is raised?

File: Lib/test/test_difflib.py [hunk 3]
```diff
@@ -273,9 +273,19 @@ def test_make_file_usascii_charset_with_nonascii_input(self):
         self.assertIn('&#305;mpl&#305;c&#305;t', output)
 
 
+class TestInputParsing(unittest.TestCase):
```
Comment (serhiy-storchaka) [inline Lib/test/test_difflib.py]: There are three existing tests for types checking: `test_mixed_types_content`, `test_mixed_types_filenames` and `test_mixed_types_dates`. I think that it is better to extract them in a separate class (they are no longer about mixing 8-bit and Unicode strings) and add a new test in this class.

File: Lib/test/test_difflib.py (hunk 2 above)
Comment (eendebakpt) [reply Lib/test/test_difflib.py]: A `TypeError`. I updated the test

File: Lib/test/test_difflib.py (hunk 3 above)
Comment (eendebakpt) [reply Lib/test/test_difflib.py]: Done. I moved the 3 existing tests to a new class and refactored the new test to be in the same style.

File: Lib/test/test_difflib.py [hunk 4]
```diff
@@ -427,58 +497,6 @@ def assertDiff(expect, actual):
                                     lineterm=b'')
         assertDiff(expect, actual)
 
-    def test_mixed_types_content(self):
```
Comment (serhiy-storchaka) [inline Lib/test/test_difflib.py:450]: To minimize the diff, I would move the new test class here.

File: Lib/difflib.py [hunk 5]
```diff
@@ -1266,6 +1266,12 @@ def _check_types(a, b, *args):
     if b and not isinstance(b[0], str):
         raise TypeError('lines to compare must be str, not %s (%r)' %
                         (type(b[0]).__name__, b[0]))
+    if isinstance(a, str):
+        raise TypeError('input must be a sequence of strings, not %s (%r)' %
```
Comment (serhiy-storchaka) [inline Lib/difflib.py]: If the user pass the content of the whole file without splitting it on lines, all this (potentially megabites) will be included in the error message. This is not good. The content of the string does not add anything to clarify the bug, the type is enough.

It is better to include only type name instead of the repr also in all other error messages, but this is not related to this PR.

Comment (eendebakpt) [reply Lib/difflib.py]: I agree we should not print the entire object. I changed it for the new error messages. Let me know whether you want me to also change the other error messages in this PR.

**Review by serhiy-storchaka — APPROVED** [review]

**Comment by serhiy-storchaka** [comment]
Thank you for your contribution @eendebakpt. There were some unrelated random test failures, so I forget to merge this at its time. I should use "auto-merge" instead.
